Code/GraphMol/Resonance.cpp

```diff
@@ -0,0 +1,1355 @@
+// $Id$
+//
+//  Copyright (C) 2015 Paolo Tosco
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#include <iostream>
+#include <GraphMol/RDKitBase.h>
+#include <GraphMol/MolOps.h>
+#include "Resonance.h"
+#include <boost/functional/hash.hpp>
+
+namespace RDKit {
+  namespace ResonanceUtils {
+    // depending on the number of atoms which won't have a complete
+    // octet (nCandSlots) and the number of those which need non-bonded
+    // electrons (nTotalSlots), the number of permutations (numComb) and
+    // a binary code (v) which indicates which of the atom indices in
+    // aiVec will be octet-unsatisfied for each permutation are computed
+    void getNumCombStartV(unsigned int nCandSlots,
+      unsigned int nTotalSlots, unsigned int &numComb, unsigned int &v)
+    {
+      numComb = 1;
+      v = 0;
+      for (unsigned int i = 0; i < nCandSlots; ++i) {
+        numComb *= (nTotalSlots - i);
+        numComb /= (i + 1);
+        v |= (1 << i);
+      }
+    }
+    
+    // get the next permutation
+    void updateV(unsigned int &v)
+    {
+      unsigned int t = (v | (v - 1)) + 1;
+      v = t | ((((t & (~t + 1)) / (v & (~v + 1))) >> 1) - 1);
+    }
+
+    // sanitize the resonance structure which has been assembled
+    void sanitizeMol(RWMol &mol)
+    {
+      unsigned int opFailed;
+      MolOps::sanitizeMol(mol, opFailed, MolOps::SANITIZE_FINDRADICALS
+        | MolOps::SANITIZE_SETAROMATICITY | MolOps::SANITIZE_ADJUSTHS);
+      if (opFailed != MolOps::SANITIZE_NONE) {
+        std::string s;
+        switch (opFailed) {
+          case MolOps::SANITIZE_FINDRADICALS:
+            s = "SANITIZE_FINDRADICALS";
+            break;
+          case MolOps::SANITIZE_SETAROMATICITY:
+            s = "SANITIZE_SETAROMATICITY";
+            break;
+          case MolOps::SANITIZE_ADJUSTHS:
+            s = "SANITIZE_ADJUSTHS";
+            break;
+        }
+        s += " failed";
+        throw std::runtime_error(s);
```

Comment:
I'm not sure I understand the logic here. Why not just let the underlying sanitization code throw an exception?